src/traces/image/defaults.js

```diff
@@ -11,13 +11,19 @@
 var Lib = require('../../lib');
 var attributes = require('./attributes');
 var constants = require('./constants');
+var dataUri = require('../../snapshot/helpers').IMAGE_URL_PREFIX;
 
 module.exports = function supplyDefaults(traceIn, traceOut) {
     function coerce(attr, dflt) {
         return Lib.coerce(traceIn, traceOut, attributes, attr, dflt);
     }
+    coerce('source');
+    if(traceOut.source && !traceOut.source.match(dataUri)) delete traceOut.source;
+    traceOut._isFromSource = !!traceOut.source;
+
     var z = coerce('z');
-    if(z === undefined || !z.length || !z[0] || !z[0].length) {
+    traceOut._isFromZ = !(z === undefined || !z.length || !z[0] || !z[0].length);
+    if(!traceOut._isFromZ && !traceOut._isFromSource) {
```

Comment:
Non-blocking: 
I suggest we rename these properties to `._hasZ` and `._hasSource`.

Reply:
Done in 3c313dbe131a01df898629b6e3bbdc4009893d7e